subsys/storage/flash_buffered_write/fbw.c

```diff
@@ -0,0 +1,210 @@
+		rc = flash_get_page_info_by_offs(flash_dev, off, &page);
+		if (rc == 0) {
+			sector->fs_off = page.start_offset;
+			sector->fs_size = page.size;
+		}
+	}
+
+	return rc;
+}
+
+int fbw_progressive_erase(struct fbw_ctx *ctx, off_t off)
+{
+	struct flash_sector sector;
+	int rc;
+
+	rc = flash_sector_from_off(ctx->flash_area, off, &sector);
+	if (rc) {
+		LOG_ERR("Unable to determine flash sector size");
+	} else {
+		if (ctx->off_last != sector.fs_off) {
+			ctx->off_last = sector.fs_off;
+			LOG_INF("Erasing sector at offset 0x%08lx",
+				(long)sector.fs_off);
+			rc = flash_area_erase(ctx->flash_area,
+					      sector.fs_off -
+					      ctx->flash_area->fa_off,
+					      sector.fs_size);
+			if (rc) {
+				LOG_ERR("Error %d while erasing sector", rc);
+			}
+		}
+	}
+
+	return rc;
+}
+
+#endif /* CONFIG_FBW_ERASE_PROGRESSIVELY */
+
+static int flash_sync(struct fbw_ctx *ctx)
+{
+	int rc = 0;
+
+	if (ctx->buf_bytes < ctx->buf_len) {
+		(void)memset(ctx->buf + ctx->buf_bytes, 0xFF,
+			     ctx->buf_len - ctx->buf_bytes);
+	}
+
+#ifdef CONFIG_FBW_ERASE_PROGRESSIVELY
+	rc = fbw_progressive_erase(ctx, ctx->bytes_written +
+				     ctx->buf_len);
+	if (rc) {
+		LOG_ERR("fbw_progressive_erase error %d offset=0x%08zx", rc,
+			ctx->bytes_written);
+		return rc;
+	}
+#endif
+
+	rc = flash_area_write(ctx->flash_area, ctx->bytes_written, ctx->buf,
+			      ctx->buf_len);
+	if (rc) {
+		LOG_ERR("flash_write error %d offset=0x%08zx", rc,
+			ctx->bytes_written);
+		return rc;
+	}
+
+	if (!flash_verify(ctx->flash_area, ctx->bytes_written, ctx->buf,
+			  ctx->buf_len)) {
+		return -EIO;
+	}
+
+	ctx->bytes_written += ctx->buf_bytes;
+	ctx->buf_bytes = 0U;
+
+	return rc;
+}
+
+int fbw_write(struct fbw_ctx *ctx, u8_t *data, size_t len, bool flush)
+{
+	int processed = 0;
+	int rc = 0;
+	int buf_empty_bytes;
+
+	while ((len - processed) >=
+	       (buf_empty_bytes = ctx->buf_len - ctx->buf_bytes)) {
+		memcpy(ctx->buf + ctx->buf_bytes, data + processed,
+		       buf_empty_bytes);
+
+		ctx->buf_bytes = ctx->buf_len;
+		rc = flash_sync(ctx);
+
+		if (rc) {
+			return rc;
+		}
+
+		processed += buf_empty_bytes;
+	}
+
+	/* place rest of the data into ctx->buf */
+	if (processed < len) {
+		memcpy(ctx->buf + ctx->buf_bytes,
+		       data + processed, len - processed);
+		ctx->buf_bytes += len - processed;
+	}
+
+	if (!flush) {
+		return rc;
+	}
+
+	if (ctx->buf_bytes > 0) {
+		/* pad the rest of ctx->buf and write it out */
+		rc = flash_sync(ctx);
+
+		if (rc) {
+			return rc;
+		}
+	}
+
+	flash_area_close(ctx->flash_area);
+	ctx->flash_area = NULL;
+
+	return rc;
+}
+
+size_t fbw_bytes_written(struct fbw_ctx *ctx)
+{
+	return ctx->bytes_written;
+}
+
+int fbw_init(struct fbw_ctx *ctx, u8_t *buf, size_t buf_len, u8_t area_id,
+	     size_t offset)
+{
+	ctx->buf = buf;
+	ctx->buf_len = buf_len;
+	ctx->bytes_written = 0;
+	ctx->buf_bytes = 0U;
+#ifdef CONFIG_FBW_ERASE_PROGRESSIVELY
+	ctx->off_last = -1;
+#endif
+
+	return flash_area_open(area_id,
```

Comment:
Can you explain the decision to move the flash_area_open() / flash_area_close() calls into fbw.c?

Naively, I would have guessed that this would be the responsibility of any callers to fbw_write() or other I/O routines. That is, the fbw.h API would just assume any flash it's asked to do I/O on would be ready for it.

I'm also a bit surprised not to see a `struct device*` in the flash context structure and the flash API being used to do the actual erasing/writing. Is there a reason to keep using flash areas here?

I ask because I've honestly always had my doubts about flash_map.h. It was basically copy/pasted into Zephyr from Mynewt code, and I didn't think it was a good idea at the time, either, since it doesn't play natively with the Zephyr driver model.

Reply:
TBH I didn't consider the rationale behind the already made choices for this module, I simply copy-pasted the functionality found in flash_img.c to allow for more re-use.

The user of this module (of which we have identified 3 until now) all have the same use case, pass a pointer to some flash area and a length, and write it to a different part of flash, internal or external.

The interface provided by this module is sufficient to solve those use cases, while requiring a very small amount of boilerplate code (opening closing flash areas etc.)

Reply:
Maybe good idea would be to transfer flash_area by the user - this will make this API the most flexible. Flash area might be then defined either in the build-time or be even created in the run-time.

Reply:
I agree

Reply:
I don't think we should stick to the current flash area API when we generalize this, to be honest.

The current `<storage/flash_map.h>` API has unnecessary inefficiencies in all of its I/O operations. The read, write, etc. routines all call device_get_binding() for every I/O op, which in the worst case is iterating through all of the devices in the system.

It also hard-codes its device IDs (`SOC_FLASH_0_ID` and `SPI_FLASH_0_ID`) instead of just using the Zephyr driver model the way it was intended with struct devices.

I would be against adding a new top-level API that relies on flash map for this reason.

Reply:
I'm working on a variant now which has the `struct device *fdev` member in the `fbw_ctx` instead, this will not use the flash_map at all, but the API of `flash_img.c` will not change, so the `flash_map` handling will be moved there as you proposed somewhere else.

Reply:
@mbolivar - I think it is easy to improve flash_area API implementation regards this, for instance inserting the driver reference into flash_area struct... (struct need to be runtime then...).